### PR TITLE
React component conversion(class to functional) - 1

### LINKED ISSUE #2709: Changing class component to Functional Component.
### Increasing Access

It is important to update components according to new Functional components.

### Feature enhancement details

The component uses useRef for setting a reference to the Codemirror instance which can be used in any useEffect and function. I am referencing a pull request for these issue. I have done a pr before and changed my commit but it was unable to reach to maintainers.

### PR DESCRIPTION
**Pull request by nahbee10**
Fixes #2709 

Changes:
- converted `IDE/components/Editor/index.jsx` to use React functional component

I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [x] has no test errors (`npm run test`)
* [x] is from a uniquely-named feature branch and is up to date with the  `develop` branch.
* [x] is descriptively named and links to an issue number, i.e. `Fixes #123`

### CODE REVIEW

File: client/modules/IDE/components/Editor/index.jsx [hunk 1]
```diff
@@ -146,168 +336,135 @@ class Editor extends React.Component {
       }
     });
 
-    this.hinter = new Fuse(hinter.p5Hinter, {
```
Comment (khanniie) [inline client/modules/IDE/components/Editor/index.jsx:149]: did we lose this in the conversion? do we not need it anymore?

Comment (khanniie) [reply client/modules/IDE/components/Editor/index.jsx:149]: if we don't need it anymore, let's also remove it from the dependencies~

File: client/modules/IDE/components/Editor/index.jsx [hunk 2]
```diff
@@ -146,168 +336,135 @@ class Editor extends React.Component {
...
+    cm.setOption('extraKeys', {
+      Tab: (_cmInstance) => {
+        if (!_cmInstance.execCommand('emmetExpandAbbreviation')) return;
+        const selection = _cmInstance.doc.getSelection();
         if (selection.length > 0) {
-          cm.execCommand('indentMore');
+          _cmInstance.execCommand('indentMore');
         } else {
-          cm.replaceSelection(' '.repeat(INDENTATION_AMOUNT));
+          _cmInstance.replaceSelection(' '.repeat(INDENTATION_AMOUNT));
```
Comment (khanniie) [inline client/modules/IDE/components/Editor/index.jsx:351]: not sure why eslint didn't catch this but i don't think this codebase uses underscores for private variables! looks like they [are following](https://github.com/processing/p5.js-web-editor/blob/develop/contributor_docs/development.md) the airbnb style guide: https://github.com/airbnb/javascript?tab=readme-ov-file#naming--leading-underscore 

this comment applies to _cmInstance but also all others with underscores

I do see that the original file had underscores too so i guess it's up to us if we want to re-align the code with the style guide named in the contributor docs (my perference) or not change things up too much : P @raclim any thoughts?

File: client/modules/IDE/components/Editor/index.jsx [hunk 3]
```diff
@@ -146,168 +336,135 @@ class Editor extends React.Component {
...
       [`${metaKey}-G`]: 'findPersistentNext',
       [`Shift-${metaKey}-G`]: 'findPersistentPrev',
       [replaceCommand]: 'replace',
-      // Cassie Tarakajian: If you don't set a default color, then when you
-      // choose a color, it deletes characters inline. This is a
-      // hack to prevent that.
-      [`${metaKey}-K`]: (cm, event) =>
-        cm.state.colorpicker.popup_color_picker({ length: 0 }),
-      [`${metaKey}-.`]: 'toggleComment' // Note: most adblockers use the shortcut ctrl+.
+      [`${metaKey}-K`]: (_cmInstMetaKey, event) =>
```
Comment (khanniie) [inline client/modules/IDE/components/Editor/index.jsx:363]: can we keep the comments from cassie and other notes? will probably be nice for the future : )

**Comment by khanniie** [comment]
Left a few comments but overall looks great! : D thanks nahee!

**Comment by khanniie** [comment]
When I test the staging link, something kind of crazy happens when i hit tidy code :O 
Screenshot 2024-08-08 at 11 26 28 AM

**Comment by khanniie** [comment]
I think we may have lost the error highlighting as well -- left is this branch and right is what's live
Screenshot 2024-08-08 at 11 28 20 AM

File: client/modules/IDE/components/Editor/index.jsx (hunk 2 above)
Comment (raclim) [reply client/modules/IDE/components/Editor/index.jsx:351]: I think [this rule is turned off in our eslint configuration](https://github.com/processing/p5.js-web-editor/blob/87b41992ccd39b0ee79b01dc886c5158ce7fdfd4/.eslintrc#L23) 🤔 but it makes sense to me to aim to align with the style guide as much as possible!

**Comment by raclim** [comment]
Just to add, I'm on the staging environment and am having trouble switching between different files in the left-hand navigation as well! 


https://github.com/user-attachments/assets/636f4ea2-35d1-4ffe-acb4-75200942396f

**Comment by raclim** [comment]
To help with organization, I'm going to close this PR as a majority of this work is being handled here is also in the draft PR for the CodeMirror upgrade. Please feel free to let me know if I missed anything and this should be reopened, thanks!

**Pull request closed without merging**